src/storage/mod.rs

```diff
@@ -568,62 +488,187 @@ impl Storage {
+                        !ctx.get_not_fill_cache(),
+                    );
+                    let result = snap_store
+                        .batch_get(&keys, &mut statistics)
+                        // map storage::txn::Error -> storage::Error
+                        .map_err(Error::from)
+                        .map(|results| results
+                            .into_iter()
+                            .zip(keys)
+                            .filter(|&(ref v, ref _k)|
+                                !(v.is_ok() && v.as_ref().unwrap().is_none())
+                            )
+                            .map(|(v, k)| match v {
+                                Ok(Some(x)) => Ok((k.raw().unwrap(), x)),
+                                Err(e) => Err(Error::from(e)),
+                                _ => unreachable!(),
+                            })
+                            .collect()
+                        )
+                        .map(|r: Vec<Result<KvPair>>| {
+                            thread_ctx.collect_key_reads(CMD, r.len() as u64);
+                            r
+                        });
+
+                    thread_ctx.collect_scan_count(CMD, &statistics);
+                    thread_ctx.collect_read_flow(ctx.get_region_id(), &statistics);
+
+                    result
+                })
+        }) {
+            Ok(val) => box val,
+            Err(_) => box future::err(Error::SchedTooBusy),
+        }
     }
 
+    /// Scan a range starting with `start_key` up to `limit` rows from the snapshot.
     pub fn async_scan(
         &self,
         ctx: Context,
         start_key: Key,
         limit: usize,
         start_ts: u64,
         options: Options,
-        callback: Callback<Vec<Result<KvPair>>>,
-    ) -> Result<()> {
-        let cmd = Command::Scan {
-            ctx: ctx,
-            start_key: start_key,
-            limit: limit,
-            start_ts: start_ts,
-            options: options,
-        };
-        let tag = cmd.tag();
-        self.schedule(cmd, StorageCb::KvPairs(callback))?;
-        KV_COMMAND_COUNTER_VEC.with_label_values(&[tag]).inc();
-        Ok(())
+    ) -> Box<Future<Item = Vec<Result<KvPair>>, Error = Error> + Send> {
+        static CMD: &'static str = "scan";
+        let engine = self.get_engine();
+        let priority = readpool::Priority::from(ctx.get_priority());
+
+        match self.read_pool.future_execute(priority, move |ctxd| {
+            let _timer = {
+                let ctxd = ctxd.clone();
+                let mut thread_ctx = ctxd.current_thread_context_mut();
+                thread_ctx.collect_command_count(CMD, priority, false);
+                thread_ctx.collect_command_duration(CMD)
+            };
+
+            engine
+                .future_snapshot(&ctx)
+                // map storage::engine::Error -> storage::txn::Error -> storage::Error
+                .map_err(txn::Error::from)
+                .map_err(Error::from)
```

Comment:
Could we define a function for L636-639 since it appears in every get operation?

Reply:
Ok